File: backend-rb/lib/domain/money.rb
# frozen_string_literal: true

module Domain
  # Money, carried as an integer count of the smallest unit plus the scale it
  # holds.
  #
  # BigDecimal will not do: BigDecimal("1250.00").to_s("F") is "1250.0", and the
  # API has to put 1250.00 on the wire because the other four backends do.
  # Arithmetic follows the same rule their decimal libraries use -- an addition
  # keeps the wider of the two scales.
  class Money
    include Comparable

    SCALE = 2

    # The bounds every [Range(typeof(decimal), ...)] in the DTOs uses.
    MIN_POSITIVE = "0.01"
    MIN_ZERO = "0.00"
    MAX = "999999999999.99"

    LITERAL = /\A[+-]?\d+(\.\d+)?\z/

    attr_reader :units, :scale

    def initialize(units, scale)
      @units = units
      @scale = scale
      freeze
    end

    class << self
      def zero = new(0, 0)

      # Reads a decimal literal, keeping the scale it was written with.
      def parse(text)
        trimmed = text.to_s.strip
        return nil unless LITERAL.match?(trimmed)

        negative = trimmed.start_with?("-")
        digits = trimmed.sub(/\A[+-]/, "")
        whole, fraction = digits.split(".")
        fraction ||= ""

        units = "#{whole}#{fraction}".to_i
        new(negative ? -units : units, fraction.length)
      end

      # Reads whatever JSON produced: a number, or a quoted number.
      def from_json(value)
        case value
        when ::Integer then new(value, 0)
        when ::Float, ::BigDecimal then parse(format("%.10f", value).sub(/0+\z/, "").sub(/\.\z/, ""))
        when ::String then parse(value)
        else nil # null, a bool, an object -- not a number the caller can use
        end
      end

      # Parses one of the bound constants above, which are always well-formed.
      def bound(text) = parse(text) || raise(ArgumentError, "money: #{text} is not a decimal")
    end

    def add(other)
      left, right, scale = align(other)
      self.class.new(left + right, scale)
    end

    def subtract(other)
      left, right, scale = align(other)
      self.class.new(left - right, scale)
    end

    def negate = self.class.new(-units, scale)

    def <=>(other)
      left, right, = align(other)
      left <=> right
    end

    def zero? = units.zero?

    # Rounds to two places, half away from zero -- and never lengthens the
    # scale, so 10 stays 10 rather than becoming 10.00.
    def round_money
      return self if scale <= SCALE

      divisor = 10**(scale - SCALE)
      magnitude = units.abs
      quotient, remainder = magnitude.divmod(divisor)
      rounded = remainder * 2 >= divisor ? quotient + 1 : quotient

      self.class.new(units.negative? ? -rounded : rounded, SCALE)
    end

    # Drops trailing fractional zeros. Used only on the savings rate.
    def trim
      value = units
      places = scale
      while places.positive? && (value % 10).zero?
        value /= 10
        places -= 1
      end
      self.class.new(value, places)
    end

    # Divides to a fixed number of places, rounding half away from zero. Money
    # has no general division because the API needs it in exactly one place.
    def divide(other, places)
      left, right, = align(other)
      return self.class.zero if right.zero?

      negative = left.negative? ^ right.negative?
      scaled = left.abs * (10**places)
      quotient, remainder = scaled.divmod(right.abs)
      rounded = remainder * 2 >= right.abs ? quotient + 1 : quotient

      self.class.new(negative ? -rounded : rounded, places)
    end

    # Renders as a bare decimal literal, keeping its scale.
    def to_s
      return units.to_s if scale.zero?

      digits = units.abs.to_s.rjust(scale + 1, "0")
      whole = digits[0...-scale]
      fraction = digits[-scale..]
      "#{units.negative? ? '-' : ''}#{whole}.#{fraction}"
    end

    # Exactly two places, whatever the scale. The CSV column wants this.
    def to_fixed2
      return to_s if scale == SCALE
      return rescale(SCALE).to_s if scale < SCALE

      round_money.to_s
    end

    def as_json(*) = self
    def inspect = "#<Money #{self}>"

    private

    def align(other)
      common = [ scale, other.scale ].max
      [ units * 10**(common - scale), other.units * 10**(common - other.scale), common ]
    end

    def rescale(places) = self.class.new(units * 10**(places - scale), places)
  end

  MONEY_ZERO = Money.zero
end

module Domain
  # Tag packing for the TagsRaw column.
  module Tags
    # US, the unit separator: it cannot occur inside a tag. Built from its
    # codepoint rather than written literally, because a raw control byte in
    # source is invisible in review and easy to mangle when copied.
    SEPARATOR = 31.chr(Encoding::UTF_8)

    module_function

    # Trims, drops blanks, and packs into the storage column.
    def join(tags)
      Array(tags).map { |tag| tag.to_s.strip }.reject(&:empty?).join(SEPARATOR)
    end

    # Unpacks the storage column. Always an array, never nil.
    def split(raw) = raw.to_s.split(SEPARATOR).reject(&:empty?)

    # Whitespace-only collapses to nil, as the other backends do for Notes.
    def trimmed_or_nil(value)
      text = value.to_s.strip
      text.empty? ? nil : text
    end
  end
end

File: backend-node/src/domain/money.ts
/**
 * Money, tags and the wire formats for dates.
 *
 * JavaScript has no decimal type, and the usual libraries normalise trailing
 * zeros away -- `new Decimal('1250.00').toString()` is `'1250'`. This API has
 * to put `1250.00` on the wire, because the .NET, Go and Python backends do.
 * So money is a pair here: an integer count of the smallest unit, plus the
 * scale it is carrying. Arithmetic follows the same rule the other backends'
 * decimal libraries use -- an addition keeps the wider of the two scales.
 */

/** The bounds every [Range(typeof(decimal), ...)] in the DTOs uses. */
export const MONEY_MIN_POSITIVE = '0.01'
export const MONEY_MIN_ZERO = '0.00'
export const MONEY_MAX = '999999999999.99'

export const MONEY_SCALE = 2

/** The separator the TagsRaw column uses; it cannot occur inside a tag. */
const TAG_SEPARATOR = '\u001F'

const TEN = 10n

function pow10(exponent: number): bigint {
  return TEN ** BigInt(exponent)
}

export class Money {
  /** The value as an integer count of 10^-scale units. */
  readonly units: bigint
  /** How many decimal places this value is carrying. Never negative. */
  readonly scale: number

  private constructor(units: bigint, scale: number) {
    this.units = units
    this.scale = scale
  }

  static of(units: bigint, scale: number): Money {
    return new Money(units, scale)
  }

  static zero(): Money {
    return new Money(0n, 0)
  }

  /** Reads a decimal literal, keeping the scale it was written with. */
  static parse(text: string): Money | null {
    const trimmed = text.trim()
    // Deliberately strict: no exponents, no separators, no currency symbols.
    if (!/^[+-]?\d+(\.\d+)?$/.test(trimmed)) return null

    const negative = trimmed.startsWith('-')
    const digits = trimmed.replace(/^[+-]/, '')
    const [whole = '', fraction = ''] = digits.split('.')

    const units = BigInt(whole + fraction)
    return new Money(negative ? -units : units, fraction.length)
  }

  /** Reads whatever JSON produced: a number, or a quoted number. */
  static fromJson(value: unknown): Money | null {
    if (typeof value === 'number') {
      // Number->string is exact for every value the API accepts, and keeps the
      // scale the caller actually wrote (12.5 stays one place).
      return Number.isFinite(value) ? Money.parse(String(value)) : null
    }
    if (typeof value === 'string') return Money.parse(value)
    return null
  }

  private align(other: Money): [bigint, bigint, number] {
    const scale = Math.max(this.scale, other.scale)
    return [
      this.units * pow10(scale - this.scale),
      other.units * pow10(scale - other.scale),
      scale,
    ]
  }

  add(other: Money): Money {
    const [left, right, scale] = this.align(other)
    return new Money(left + right, scale)
  }

  subtract(other: Money): Money {
    const [left, right, scale] = this.align(other)
    return new Money(left - right, scale)
  }

  negate(): Money {
    return new Money(-this.units, this.scale)
  }

  compare(other: Money): number {
    const [left, right] = this.align(other)
    if (left < right) return -1
    return left > right ? 1 : 0
  }

  lessThan(other: Money): boolean {
    return this.compare(other) < 0
  }

  greaterThan(other: Money): boolean {
    return this.compare(other) > 0
  }

  isZero(): boolean {
    return this.units === 0n
  }

  /**
   * Rounds to two places, half away from zero -- and never lengthens the
   * scale, so 10 stays 10 rather than becoming 10.00.
   */
  round(): Money {
    if (this.scale <= MONEY_SCALE) return this
    const divisor = pow10(this.scale - MONEY_SCALE)
    const negative = this.units < 0n
    const magnitude = negative ? -this.units : this.units

    const quotient = magnitude / divisor
    const remainder = magnitude % divisor
    // Half away from zero: a remainder of exactly half rounds up.
    const rounded = remainder * 2n >= divisor ? quotient + 1n : quotient

    return new Money(negative ? -rounded : rounded, MONEY_SCALE)
  }

  /** Drops trailing fractional zeros. Used only on the savings rate. */
  trim(): Money {
    let { units, scale } = this
    while (scale > 0 && units % TEN === 0n) {
      units /= TEN
      scale -= 1
    }
    return new Money(units, scale)
  }

  /** Renders as a bare decimal literal, keeping its scale. */
  toString(): string {
    if (this.scale === 0) return this.units.toString()

    const negative = this.units < 0n
    const digits = (negative ? -this.units : this.units).toString().padStart(this.scale + 1, '0')
    const whole = digits.slice(0, digits.length - this.scale)
    const fraction = digits.slice(digits.length - this.scale)
    return `${negative ? '-' : ''}${whole}.${fraction}`
  }

  /** Exactly two places, whatever the scale. The CSV column wants this. */
  toFixed2(): string {
    return this.scale === MONEY_SCALE ? this.toString() : this.rescale(MONEY_SCALE).toString()
  }

  private rescale(scale: number): Money {
    if (scale === this.scale) return this
    if (scale > this.scale) return new Money(this.units * pow10(scale - this.scale), scale)
    return this.round()
  }
}

export const MONEY_ZERO = Money.zero()

/** Parses a bound like "0.01"; the constants above are always well-formed. */
export function bound(text: string): Money {
  const parsed = Money.parse(text)
  if (parsed === null) throw new Error(`money: ${text} is not a decimal`)
  return parsed
}

// --- tags -------------------------------------------------------------------

/** Trims, drops blanks, and packs into the storage column. */
export function joinTags(tags: readonly string[] | null | undefined): string {
  return (tags ?? [])
    .map((tag) => tag.trim())
    .filter((tag) => tag.length > 0)
    .join(TAG_SEPARATOR)
}

/** Unpacks the storage column. Always an array, never null. */
export function splitTags(raw: string | null | undefined): string[] {
  if (!raw) return []
  return raw.split(TAG_SEPARATOR).filter((tag) => tag.length > 0)
}

/** Whitespace-only collapses to null, as the other backends do for Notes. */
export function trimmedOrNull(value: string | null | undefined): string | null {
  const text = (value ?? '').trim()
  return text.length > 0 ? text : null
}
